File: frontend/lib/api/simulation.ts
import { CircuitSchema, SimulationResultSchema } from "../types/quantum";

const BACKEND_BASE_URL =
  process.env.NEXT_PUBLIC_BACKEND_URL || "http://localhost:8000";

export class SimulationApiError extends Error {
  statusCode: number;
  detail: string;

  constructor(message: string, statusCode: number, detail: string) {
    super(message);
    this.name = "SimulationApiError";
    this.statusCode = statusCode;
    this.detail = detail;
  }
}

/**
 * Executes quantum circuit simulation via FastAPI backend and Qiskit Aer.
 * 
 * @param circuit - Standardized circuit schema payload
 * @returns Structured simulation results with counts and basis probabilities
 * @throws SimulationApiError on validation or execution failures
 */
export async function runSimulation(
  circuit: CircuitSchema
): Promise<SimulationResultSchema> {
  const url = `${BACKEND_BASE_URL}/api/simulation/run`;

  try {
    const response = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(circuit),
    });

    if (!response.ok) {
      let detail = `Simulation request failed with status ${response.status}`;
      try {
        const errorData = (await response.json()) as Record<string, unknown>;
        if (errorData && typeof errorData.detail === "string") {
          detail = errorData.detail;
        } else if (Array.isArray(errorData?.detail)) {
          detail = (errorData.detail as Array<{ msg?: string }>)
            .map((e) => e.msg || JSON.stringify(e))
            .join("; ");
        }
      } catch {
        // Response body was not JSON
      }
      throw new SimulationApiError(detail, response.status, detail);
    }

    const data: SimulationResultSchema = await response.json();
    return data;
  } catch (error: unknown) {
    if (error instanceof SimulationApiError) {
      throw error;
    }
    // Network connectivity error
    throw new SimulationApiError(
      "Quantum simulator is unavailable. Make sure the FastAPI server is running at http://localhost:8000.",
      0,
      "Backend connection error: please ensure FastAPI server is running on port 8000."
    );
  }
}

/**
 * Checks backend health status.
 */
export async function checkBackendHealth(): Promise<boolean> {
  try {
    const res = await fetch(`${BACKEND_BASE_URL}/health`, {
      method: "GET",
      cache: "no-store",
    });
    if (res.ok) {
      const data = await res.json();
      return data?.status === "ok";
    }
    return false;
  } catch {
    return false;
  }
}

File: backend/app/api/simulation.py
import logging
from fastapi import APIRouter, Depends, HTTPException, status
from app.schemas.circuit import CircuitSchema
from app.schemas.simulation import SimulationResultSchema
from app.services.quantum.simulator import QuantumSimulationService

logger = logging.getLogger(__name__)

router = APIRouter()


def get_simulation_service() -> QuantumSimulationService:
    """Dependency provider for the quantum simulation service."""
    return QuantumSimulationService()


@router.post(
    "/run",
    response_model=SimulationResultSchema,
    status_code=status.HTTP_200_OK,
    summary="Run quantum circuit simulation",
    description="Accepts standardized Circuit JSON, executes simulation via Qiskit Aer, and returns counts and probabilities."
)
async def run_simulation(
    circuit: CircuitSchema,
    service: QuantumSimulationService = Depends(get_simulation_service)
) -> SimulationResultSchema:
    """Simulate a quantum circuit and return measurement counts and computational basis probabilities."""
    try:
        result = service.run_simulation(circuit)
        return result
    except ValueError as ve:
        logger.warning(f"Circuit validation error: {str(ve)}")
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(ve))
    except RuntimeError as re:
        logger.error(f"Simulation runtime error: {str(re)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Simulation failed to execute on the quantum backend."
        )
    except Exception as e:
        logger.exception(f"Unexpected simulation error: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="An unexpected error occurred during quantum circuit simulation."
        )
